Make run_decomp work

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129547
* __->__ #129249
* #129092
* #128077


In this PR, we implement the first version of training_ir.run_decomp functionality. Since we don't return the modified buffers as extra output in training IR, our previous strategy of reusing graph signature won't work. In fact, this run_decomp is more similar to retracing. So i reuse some of export steps here. After this PR:
export_for_training().run_decomp({}, _preserve_ops=[all 183 ops]) == export_for_predispatch() - autograd_manipulating_ops. 

Differential Revision: [D59069090](https://our.internmc.facebook.com/intern/diff/D59069090)

tugsbayasgalan committed:
Make run_decomp work

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

zhxchen17 commented on torch/export/exported_program.py:
```diff
@@ -856,6 +946,14 @@ def run_decompositions(
         """
         from torch._decomp import core_aten_decompositions
 
+        def _get_placeholders(gm):
+            placeholders = []
+            for node in gm.graph.nodes:
+                if node.op != "placeholder":
+                    break
+                placeholders.append(node)
+            return placeholders
+
```
@tugsbayasgalan why we need this?

tugsbayasgalan replied:
Not needed. Good catch!

tugsbayasgalan committed:
Update on "Make run_decomp work"




[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59069090).

tugsbayasgalan committed:
Update on "Make run_decomp work"


Differential Revision: [D59069090](https://our.internmc.facebook.com/intern/diff/D59069090)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make run_decomp work"



In this PR, we implement the first version of training_ir.run_decomp functionality. Since we don't return the modified buffers as extra output in training IR, our previous strategy of reusing graph signature won't work. In fact, this run_decomp is more similar to retracing. So i reuse some of export steps here. After this PR:
export_for_training().run_decomp({}, _preserve_ops=[all 183 ops]) == export_for_predispatch() - autograd_manipulating_ops. 

Differential Revision: [D59069090](https://our.internmc.facebook.com/intern/diff/D59069090)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59069090).
